ledger/ledger-api-common/src/test/suite/scala/com/digitalasset/platform/server/api/services/grpc/GrpcHealthServiceSpec.scala

```diff
@@ -31,24 +29,253 @@ final class GrpcHealthServiceSpec
+      val service = new GrpcHealthService(new HealthChecks("component" -> unhealthyComponent))
+
+      val response =
+        Await.result(service.check(serviceRequestFor("component")), patienceConfig.timeout)
+
+      response should be(notServingResponse)
+    }
+
+    "report SERVING when querying a healthy component alongside other, unhealthy components" in {
+      val service = new GrpcHealthService(
+        new HealthChecks(
+          "component A" -> healthyComponent,
+          "component B" -> healthyComponent,
+          "component C" -> unhealthyComponent,
+        ))
+
+      val response =
+        Await.result(service.check(serviceRequestFor("component B")), patienceConfig.timeout)
+
+      response should be(servingResponse)
+    }
+
+    "report NOT_SERVING when querying an unhealthy component alongside other, healthy components" in {
+      val service = new GrpcHealthService(
+        new HealthChecks(
+          "component A" -> unhealthyComponent,
+          "component B" -> healthyComponent,
+          "component C" -> healthyComponent,
+        ))
+
+      val response =
+        Await.result(service.check(serviceRequestFor("component A")), patienceConfig.timeout)
+
+      response should be(notServingResponse)
+    }
+
     "observe changes in health" in {
       val responseObserver = new MockServerCallStreamObserver[HealthCheckResponse]
-      val service = new GrpcHealthService()
 
-      service.watch(request, responseObserver)
-      responseObserver.demandResponse()
+      var componentAHealth: HealthStatus = Healthy
+      var componentBHealth: HealthStatus = Healthy
+      var componentCHealth: HealthStatus = Healthy
+      val service = new GrpcHealthService(
+        new HealthChecks(
+          "component A" -> new StubReporter(componentAHealth),
+          "component B" -> new StubReporter(componentBHealth),
+          "component C" -> new StubReporter(componentCHealth),
+        ),
+        maximumWatchFrequency = 1.millisecond,
+      )
+
+      service.watch(allServicesRequest, responseObserver)
+      responseObserver.demandResponse(count = 5)
 
       eventually {
         responseObserver.elements should be(Vector(servingResponse))
       }
+
+      componentBHealth = Unhealthy
+      eventually {
+        responseObserver.elements should be(Vector(servingResponse, notServingResponse))
+      }
+
+      componentBHealth = Healthy
+      eventually {
+        responseObserver.elements should be(
+          Vector(servingResponse, notServingResponse, servingResponse))
+      }
+
+      componentAHealth = Unhealthy
+      eventually {
+        responseObserver.elements should be(
+          Vector(servingResponse, notServingResponse, servingResponse, notServingResponse))
+      }
+
+      componentCHealth = Unhealthy
+      eventually {
+        responseObserver.elements should be(
+          Vector(servingResponse, notServingResponse, servingResponse, notServingResponse))
```

Comment:
Please add a comment that setting `Component C` to unhealthy doesn't produce a new response because due to status didn't change. Naively (due to previous exposure to home-brewed health checking systems) I thought that a new `notServing` would be emitted. (oh, and because up to now the rhythm of the test was do add another response).

Reply:
Done.